resin/logs.py

```diff
@@ -1,133 +1,194 @@
+    """
+    This is low level class and is not meant to be used by end users directly.
+    """
 
+    def getContext(self, hostname, port):
+        return ClientContextFactory.getContext(self)
 
-def _print_deprecation_warning():
-    logging.warning(" using legacy logging services, this will stop working shortly.\nPlease update to ensure logs are correctly retrieved in future.")
 
+class StreamingParser(Protocol):
+    """
+    This is low level class and is not meant to be used by end users directly.
+    """
 
-# TODO: https://github.com/resin-io/resin-sdk/pull/277/files
-class Logs(object):
+    def __init__(self, callback, error):
+        self.callback = callback
+        self.error = error
+
+    def dataReceived(self, data):
+        d = Deferred()
+        d.addCallback(self.display_data)
+        d.addErrback(self.error_func)
+        d.callback(data)
+
+    def connectionLost(self, reason):
+        pass
+
+    def display_data(self, data):
+        lines = data.splitlines()
+        for line in lines:
+            try:
+                self.callback(json.loads(line))
+            except:
+                # Ignore incomplete log entry, this can happen with the old logs
+                pass
+
+    def error_func(self, error):
+        if self.error:
+            d = Deferred()
+            d.addErrback(self.error)
+            d.errback(DefaultException(error))
+
+
+def cbRequest(response, callback, error):
+    protocol = StreamingParser(callback, error)
+    response.deliverBody(protocol)
+    return protocol
+
+
+def cbDrop(protocol):
+    protocol.transport.stopProducing()
+    protocol.transport.loseConnection()
+
+
+class Subscription(object):
     """
-    This class implements functions that allow processing logs from device.
+    This is low level class and is not meant to be used by end users directly.
+    """
+
+    def __init__(self):
+        self.context_factory = WebClientContextFactory()
+        self.settings = Settings()
 
-    This class is implemented using pubnub python sdk.
+    def add(self, uuid, callback, error=None, count=None):
+        query = 'stream=1'
+        if count:
+            query = 'stream=1&count={}'.format(count)
 
-    For more details about pubnub, please visit: https://www.pubnub.com/docs/python/pubnub-python-sdk
+        url = urljoin(
+            self.settings.get('api_endpoint'),
+            '/device/v2/{uuid}/logs?{query}'.format(uuid=uuid, query=query)
+        )
+        headers = {}
+        headers[b'Authorization'] = ['Bearer {:s}'.format(self.settings.get('token')).encode()]
 
+        agent = Agent(reactor, self.context_factory)
+        d = agent.request(b'GET', url.encode(), Headers(headers), None)
+        d.addCallback(cbRequest, callback, error)
+        return d
+
+    def run(self):
+        if not reactor.running:
+            Thread(target=reactor.run, args=(False,)).start()
+
+    def stop(self, d):
+        d.addCallback(cbDrop)
```

Comment:
Will this be called from the reactor thread? It's unsafe if it's not

Reply:
it will be called from the reactor thread, if in the code, we don't spawn a new thread by doing `reactor.callinThread()` then no need to worry about this. Every call to reactor API will be run in the reactor thread.

Reply:
stop is not called by the reactor thread. it's called by the main thread

Reply:
Changed to a thread-safe method.